File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/normx/collapse_spaces.go
package normx

// CollapseSpaces normalizes with the collapse spaces rule.
func CollapseSpaces(text string) string {
	out := make([]byte, 0, len(text))
	pending := false
	for i := 0; i < len(text); i++ {
		if text[i] == ' ' {
			pending = true
			continue
		}
		if pending && len(out) > 0 {
			out = append(out, ' ')
		}
		pending = false
		out = append(out, text[i])
	}
	return string(out)
}

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/norm/CollapseSpaces.java
package com.freight.norm;

/** Normalizer: collapse spaces. */
public final class CollapseSpaces implements Normalizer {

    @Override
    public String name() {
        return "collapse_spaces";
    }

    @Override
    public String apply(String text) {
        StringBuilder out = new StringBuilder(text.length());
        boolean pending = false;
        for (int i = 0; i < text.length(); i++) {
            char c = text.charAt(i);
            if (c == ' ') {
                pending = true;
                continue;
            }
            if (pending && out.length() > 0) {
                out.append(' ');
            }
            pending = false;
            out.append(c);
        }
        return out.toString();
    }
}
